### PR TITLE
Android 10 support for X509TrustManager

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
Earlier failing test

```
08:20:04 V/InstrumentationResultParser: 1) testCustomTrustManager(okhttp.android.test.OkHttpTest)
08:20:04 V/InstrumentationResultParser: java.lang.IllegalArgumentException: Required method checkServerTrusted(X509Certificate[], String, String, String) missing
08:20:04 V/InstrumentationResultParser: at android.net.http.X509TrustManagerExtensions.<init>(X509TrustManagerExtensions.java:71)
08:20:04 V/InstrumentationResultParser: at okhttp3.internal.platform.android.Android10CertificateChainCleaner.<init>(Android10CertificateChainCleaner.kt:36)
08:20:04 V/InstrumentationResultParser: at okhttp3.internal.platform.Android10Platform.buildCertificateChainCleaner(Android10Platform.kt:62)
08:20:04 V/InstrumentationResultParser: at okhttp3.internal.tls.CertificateChainCleaner$Companion.get(CertificateChainCleaner.kt:42)
08:20:04 V/InstrumentationResultParser: at okhttp3.OkHttpClient$Builder.sslSocketFactory(OkHttpClient.kt:751)
08:20:04 V/InstrumentationResultParser: at okhttp.android.test.OkHttpTest.testCustomTrustManager(OkHttpTest.kt:437)
08:20:04 V/InstrumentationResultParser: at java.lang.reflect.Method.invoke(Native Method)
08:20:04 V/InstrumentationResultParser: at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
08:20:04 V/InstrumentationResultParser: at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
08:20:04 V/InstrumentationResultParser: at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
08:20:04 V/InstrumentationResultParser: at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
08:20:04 V/InstrumentationResultParser: at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:111)
08:20:04 V/InstrumentationResultParser: at okhttp3.testing.PlatformRule$apply$1.evaluate(PlatformRule.kt:66)
```

File: okhttp/src/main/java/okhttp3/internal/platform/android/Android10CertificateChainCleaner.kt [hunk 1]
```diff
@@ -52,4 +53,14 @@ internal class Android10CertificateChainCleaner(
           other.trustManager === this.trustManager
 
   override fun hashCode(): Int = System.identityHashCode(trustManager)
+
+  companion object {
+    fun buildIfSupported(trustManager: X509TrustManager): Android10CertificateChainCleaner? {
+      return try {
+        Android10CertificateChainCleaner(trustManager)
+      } catch (iae: IllegalArgumentException) {
+        null
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/platform/android/Android10CertificateChainCleaner.kt]: add a comment pointing at why this is a reasonable thing to do here?

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/platform/android/Android10CertificateChainCleaner.kt]: also – could potentially limit try/catch to creating the `X509TrustManagerExtensions` instance, and then pass _that_ to the `Android10CertificateChainCleaner` constructor

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 2]
```diff
@@ -416,6 +418,64 @@ class OkHttpTest {
...
+
+      override fun checkServerTrusted(p0: Array<out X509Certificate>?, p1: String?) {}
+
+      override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
+    }
+
+    val sslContext = Platform.get().newSSLContext().apply {
+      init(null, arrayOf(tm), null)
+    }
+    val sf = sslContext.socketFactory
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: Java names? trustManager, socketFactory, hostnameVerifier

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 3]
```diff
@@ -416,6 +418,64 @@ class OkHttpTest {
...
+  @Test
+  fun testCustomTrustManagerWithCheck() {
+    assumeNetwork()
+
+    val tm = object : X509TrustManager {
+      override fun checkClientTrusted(p0: Array<out X509Certificate>?, p1: String?) {}
+
+      override fun checkServerTrusted(p0: Array<out X509Certificate>?, p1: String?) {}
+
+      fun checkServerTrusted(p0: Array<out X509Certificate>?, p1: String?, p2: String?) = p0?.toList()
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: p0, etc. –> names from the overridden things?

**Comment by yschimke** [comment]
> Love the structure of this but I’ve got some nits on a bit of your style!

Happy to address, it's a bit clumsy at the moment because I was in intellij without good editing support.  I'll fix tomorrow.

**Comment by yschimke** [comment]
Follow on bug - https://github.com/square/okhttp/issues/5689

**Comment by aaravraj960899867** [comment]
Support me

**Comment by yschimke** [comment]
image

**Pull request merged**